From 56924e1c95a90c6d78c43639b4625c672c19048f Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Wed, 22 Nov 2017 21:29:41 +0900
Subject: [PATCH 377/909] media: i2c: adv748x: Add virtual channel option

commit 7c8db01aa11a336a3f74d6c8f376d1394732e7f3 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

At the adv7482_txa and adv7482_txb nodes in dts file,
by specifying "virtual-channel = <0 or 1 or 2 or 3>;",
add an option that can be changed it.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 Documentation/devicetree/bindings/media/i2c/adv748x.txt | 3 +++
 drivers/media/i2c/adv748x/adv748x-csi2.c                | 9 ++++++++-
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/Documentation/devicetree/bindings/media/i2c/adv748x.txt b/Documentation/devicetree/bindings/media/i2c/adv748x.txt
index 9a74db585796..a33218b6dfd2 100644
--- a/Documentation/devicetree/bindings/media/i2c/adv748x.txt
+++ b/Documentation/devicetree/bindings/media/i2c/adv748x.txt
@@ -20,6 +20,9 @@ Optional Properties:
   - interrupts: Specify the interrupt lines for the ADV748x
   - data-lanes: Specify the data-lanes for the adv7482_txa of ADV748x as follows,
 		data-lanes = <1> or data-lanes = <1,2> data-lanes = <1,2,3,4>
+  - virtual-channel: Specify the same virtual-channel value for the both and adv7482_txa
+		     and adv7482_txb node of ADV748x as follows, If not specified, it is 0.
+		     virtual-channel = <0 or 1 or 2 or 3>;
 
 The device node must contain one 'port' child node per device input and output
 port, in accordance with the video interface bindings defined in
diff --git a/drivers/media/i2c/adv748x/adv748x-csi2.c b/drivers/media/i2c/adv748x/adv748x-csi2.c
index 820b44ed56a8..de30d489b60d 100644
--- a/drivers/media/i2c/adv748x/adv748x-csi2.c
+++ b/drivers/media/i2c/adv748x/adv748x-csi2.c
@@ -268,6 +268,7 @@ int adv748x_csi2_init(struct adv748x_state *state, struct adv748x_csi2 *tx)
 {
 	struct device_node *ep;
 	int ret;
+	unsigned int ch;
 
 	/* We can not use container_of to get back to the state with two TXs */
 	tx->state = state;
@@ -280,8 +281,14 @@ int adv748x_csi2_init(struct adv748x_state *state, struct adv748x_csi2 *tx)
 		return -ENODEV;
 	}
 
+	if (of_property_read_u32(ep, "virtual-channel", &ch))
+		ch = 0;
+
+	if (ch > 3)
+		return -EINVAL;
+
 	/* Initialise the virtual channel */
-	adv748x_csi2_set_virtual_channel(tx, 0);
+	adv748x_csi2_set_virtual_channel(tx, ch);
 
 	adv748x_subdev_init(&tx->sd, state, &adv748x_csi2_ops,
 			    MEDIA_ENT_F_UNKNOWN,
-- 
2.17.1

